// @loaders.gl/gltf/dist/es6/lib/deprecated/gltf-parser.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { TextDecoder, fetchFile } from '@loaders.gl/core';
import { getFullUri } from '../gltf-utils/gltf-utils';
import { getGLTFAccessors, getGLTFAccessor } from '../gltf-utils/gltf-attribute-utils';
import { KHR_DRACO_MESH_COMPRESSION, UBER_POINT_CLOUD_EXTENSION } from '../gltf-constants';
import GLBParser from './glb-parser';
import GLTFPostProcessorOld from './gltf-post-processor-old';
var DEFAULT_OPTIONS = {
  fetchLinkedResources: true,
  fetch: fetchFile,
  decompress: false,
  DracoLoader: null,
  postProcess: true,
  createImages: false,
  log: console
};
export default class GLTFParser {
  parse(gltf) {
    var _this = this,
        _arguments = arguments;

    return _asyncToGenerator(function* () {
      var options = _arguments.length > 1 && _arguments[1] !== undefined ? _arguments[1] : {};
      options = Object.assign({}, DEFAULT_OPTIONS, options);

      _this.parseSync(gltf, _objectSpread({}, options, {
        postProcess: false,
        decompress: false
      }));

      if (options.fetchLinkedResources) {
        yield _this._loadLinkedAssets(options);
      }

      if (options.decompress) {
        _this._decompressMeshes(options);
      }

      if (options.postProcess) {
        var postProcessor = new GLTFPostProcessorOld();
        postProcessor.postProcess(_this.gltf, _this.glbParser, options);
      }

      return _this.gltf;
    })();
  }

  parseSync(gltf) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    options = Object.assign({}, DEFAULT_OPTIONS, options);

    if (gltf instanceof ArrayBuffer && !GLBParser.isGLB(gltf, options)) {
      var textDecoder = new TextDecoder();
      gltf = textDecoder.decode(gltf);
    }

    if (typeof gltf === 'string') {
      gltf = JSON.parse(gltf);
    }

    if (gltf instanceof ArrayBuffer) {
      this.glbParser = new GLBParser();
      this.gltf = this.glbParser.parseSync(gltf).json;
      this.json = this.gltf;
    } else {
      this.glbParser = null;
      this.gltf = gltf;
      this.json = gltf;
    }

    if (options.decompress) {
      this._decompressMeshes(options);
    }

    if (options.postProcess) {
      var postProcessor = new GLTFPostProcessorOld();
      postProcessor.postProcess(this.gltf, this.glbParser, options);
    }

    return this.gltf;
  }

  getApplicationData(key) {
    var data = this.json[key];
    return data;
  }

  getExtraData(key) {
    var extras = this.json.extras || {};
    return extras[key];
  }

  getExtension(extensionName) {
    var isExtension = this.getUsedExtensions().find(name => name === extensionName);
    var extensions = this.json.extensions || {};
    return isExtension ? extensions[extensionName] || true : null;
  }

  getRequiredExtension(extensionName) {
    var isRequired = this.getRequiredExtensions().find(name => name === extensionName);
    return isRequired ? this.getExtension(extensionName) : null;
  }

  getRequiredExtensions() {
    return this.json.extensionsRequired || [];
  }

  getUsedExtensions() {
    return this.json.extensionsUsed || [];
  }

  getScene(index) {
    return this._get('scenes', index);
  }

  getNode(index) {
    return this._get('nodes', index);
  }

  getSkin(index) {
    return this._get('skins', index);
  }

  getMesh(index) {
    return this._get('meshes', index);
  }

  getMaterial(index) {
    return this._get('materials', index);
  }

  getAccessor(index) {
    return this._get('accessors', index);
  }

  getCamera(index) {
    return null;
  }

  getTexture(index) {
    return this._get('textures', index);
  }

  getSampler(index) {
    return this._get('samplers', index);
  }

  getImage(index) {
    return this._get('images', index);
  }

  getBufferView(index) {
    return this._get('bufferViews', index);
  }

  getBuffer(index) {
    return this._get('buffers', index);
  }

  _get(array, index) {
    if (typeof index === 'object') {
      return index;
    }

    var object = this.gltf[array] && this.gltf[array][index];

    if (!object) {
      console.warn("glTF file error: Could not find ".concat(array, "[").concat(index, "]"));
    }

    return object;
  }

  _loadLinkedAssets(options) {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      return yield Promise.all(_this2.gltf.buffers.map(buffer => _this2._loadBuffer(buffer, options)));
    })();
  }

  _loadBuffer(buffer, options) {
    return _asyncToGenerator(function* () {
      if (buffer.uri && options.uri) {
        var fetch = options.fetch || window.fetch;
        var uri = getFullUri(buffer.uri, options.uri);
        var response = yield fetch(uri);
        var arrayBuffer = yield response.arrayBuffer();
        buffer.data = arrayBuffer;
        buffer.uri = null;
      }
    })();
  }

  _decompressMeshes(options) {
    if (!options.DracoLoader || !options.decompress) {
      return;
    }

    for (var mesh of this.gltf.meshes || []) {
      for (var primitive of mesh.primitives) {
        this._decompressKhronosDracoPrimitive(primitive, options);

        this._decompressUberDracoPrimitive(primitive, options);

        if (!primitive.attributes || Object.keys(primitive.attributes).length === 0) {
          throw new Error('Empty glTF primitive: decompression failure?');
        }
      }
    }

    this._removeExtension(KHR_DRACO_MESH_COMPRESSION);

    this._removeExtension(UBER_POINT_CLOUD_EXTENSION);
  }

  _decompressKhronosDracoPrimitive(primitive, options) {
    var compressedMesh = primitive.extensions && primitive.extensions[KHR_DRACO_MESH_COMPRESSION];

    if (!compressedMesh) {
      return;
    }

    delete primitive.extensions[KHR_DRACO_MESH_COMPRESSION];

    var buffer = this._getBufferViewArray(compressedMesh.bufferView);

    var decodedData = options.DracoLoader.parseSync(buffer);
    primitive.attributes = getGLTFAccessors(decodedData.attributes);

    if (decodedData.indices) {
      primitive.indices = getGLTFAccessor(decodedData.indices);
    }
  }

  _decompressUberDracoPrimitive(primitive, options) {
    var compressedMesh = primitive.extensions && primitive.extensions[UBER_POINT_CLOUD_EXTENSION];

    if (!compressedMesh) {
      return;
    }

    if (primitive.mode !== 0) {
      throw new Error(UBER_POINT_CLOUD_EXTENSION);
    }

    delete primitive.extensions[UBER_POINT_CLOUD_EXTENSION];

    var buffer = this._getBufferViewArray(compressedMesh.bufferView);

    var decodedData = options.DracoLoader.parseSync(buffer);
    primitive.attributes = decodedData.attributes;
  }

  _getBufferViewArray(bufferViewIndex) {
    var bufferView = this.gltf.bufferViews[bufferViewIndex];

    if (this.glbParser) {
      return this.glbParser.getBufferView(bufferView);
    }

    var buffer = this.gltf.buffers[bufferView.buffer].data;
    var byteOffset = bufferView.byteOffset || 0;
    return new Uint8Array(buffer, byteOffset, bufferView.byteLength);
  }

  _removeExtension(extensionName) {
    if (this.json.extensionsRequired) {
      this._removeStringFromArray(this.json.extensionsRequired, extensionName);
    }

    if (this.json.extensionsUsed) {
      this._removeStringFromArray(this.json.extensionsUsed, extensionName);
    }

    if (this.json.extensions) {
      delete this.json.extensions[extensionName];
    }
  }

  _removeStringFromArray(array, string) {
    var found = true;

    while (found) {
      var index = array.indexOf(string);

      if (index > -1) {
        array.splice(index, 1);
      } else {
        found = false;
      }
    }
  }

}
